Add systemd service file

tues opened the pull request:
Adds a user service file for systemd and makes `install.py` download it
automatically to `~/.bloop/systemd/bloop.service`.
Updates the documentation with steps needed to enable this service.

Fixes systemd part of #337. A Desktop Entry solution will follow (most likely).

Duhemm commented on etc/systemd/bloop.service:
```diff
@@ -0,0 +1,8 @@
+Description=Bloop Scala build server
+
+[Service]
+ExecStart=%h/.bloop/bloop server
```
Should we assume that `bloop` is in the `$PATH` and just use `bloop server` here?

I'm asking because we support `--destination` in the installation script, which will override the directory where Bloop is installed.

rberenguel replied:
Seems better to expect it to be on the path, otherwise we need to get the installation script to template this file with the chosen location (which sounds good as a solution too)

tues replied:
True, I haven't noticed `--destination` when I started working on this. OTOH I don't think we can expect to have `bloop` in `$PATH` because systemd services don't use `.bashrc` etc., so a template seems to be the only solution. We'll also need templates for Desktop Entries anyway because we can't use environment variables at all there.

tues replied:
Alternatively, we could try to add something like `BLOOP_PATH` to `~/.config/environment.d/`, which should be visible to the service and we could then use it in place of `%h/.bloop`. However, it's not that much different from using a template for `bloop.service`.

I don't think we should change `PATH` automatically because that's something that some users might not want to do.

jvican commented on bin/install.py:
```diff
@@ -126,3 +130,6 @@ def makedir(directory):
 download_and_install(BASH_COMPLETION_URL, BASH_COMPLETION_TARGET)
 print("Installed Bash completion in '%s'" % BASH_COMPLETION_TARGET)
+
+download_and_install(SYSTEMD_SERVICE_URL, SYSTEMD_SERVICE_TARGET)
```
Can we do this if it's only Linux/BSD?

tues replied:
Is there any reason to do this only for Linux/BSD? Bash and zsh completions are downloaded even if you don't have bash/zsh installed.

jvican commented on etc/systemd/bloop.service:
```diff
@@ -0,0 +1,8 @@
+[Unit]
+Description=Bloop Scala build server
+
+[Service]
```
Can we redirect logs to a file? Maybe `/var/log` if we can or a log file called `server.log` in the bloop installation target?

tues replied:
I think the standard way to do this in `systemd` is to redirect `stdout`/`stderr` to system log and let it be handled by `journalctl`. The issue with `/var/log` is that the user running Bloop most likely won't have permissions to write there.

tues replied:
Please see https://github.com/scalacenter/bloop/pull/440/commits/e2a4ff525da6a92e95598aa04d3f5eeb98592ee6

tues commented:
I've added `download_and_install_template` to `install.py` and used it for `bloop.service` in place of normal `download_and_install`. It resulted in some duplicated code, but we don't want to download `coursier` or `nailgun` using `urlopen`, so I've kept `urlretrieve` for these.

tues commented:
> Why not urllib here? urllib2 does not have retrieve but both have open, I'd guess they do what we want the same and is more consistent

`urllib2` doesn't silently ignore HTTP errors, but yes, I guess it's better to replace it with `urllib` for consistency and worry about error handling later.

> Maybe it's better to use {bloop_installation_target} in the template and then use template.format(bloop_installation_target=BLOOP_INSTALLATION_TARGET) (takes advantage of format to do the work for us)

This is an interesting idea but it would mean that we need to remember to double any literal `{` and `}` in the templates, which I think is a bit annoying and risky. Currently we're not using these characters in `bloop.service`, but we might use them quite often in other files in the future. What do you think?

rberenguel commented:
Dang, you are right @tues , we might have issues with braces later on. It's just that I inherited quite a lot of "non-quite standard templating systems" in Python and other languages and I've found that in the long run it pays to just use format. If we name arguments the script will fail outright on creating the template if there are unmatched braces, so it should be relatively safe. Do you have any preference @jvican @Duhemm ?

tues commented:
Created a new branch for the Desktop Entry stuff: https://github.com/tues/bloop/commit/0fd7655cc9b7896e4de303107286df8a0df68811
Please let me know if you want it included as well.

valencik commented on bin/install.py:
```diff
@@ -89,6 +97,19 @@ def download_and_install(url, target):
         print("Couldn't download %s, please try again." % url)
         sys.exit(1)
 
+def download_and_install_template(url, target):
```
I think you could reuse the existing `download_and_install` and then modify the template once on disk (inside your `replace_template_variables`) instead of introducing socket handling and new urllib code

jvican replied:
This is a good idea :+1:

jvican commented:
Hey guys, I'll get back to this PR probably tomorrow in a break, I'm a little bit busy preparing our presentation for Scalasphere. Thanks for your work here @tues, it's excellent.

jvican commented:
I was fiddling with this PR, the documentation and instructions to get running are :100:. One thing I've noticed though is that I get different results if I run with python2 or python3:
  
```
jvican in /data/rw/code/scala/loop                                                                                                                                                                [15:13:41] 
> $ python2 /data/rw/code/scala/loop/bin/install.py --dest ~/.bloop --nailgun 615737a9408654744bc066fa6c3cb938f3f139c6 --version 1.0.0-M9                                     [±tues-ticket/337]
Installed bloop server in '/home/jvican/.bloop/blp-server'
Installed bloop client in '/home/jvican/.bloop/bloop'
Installed zsh completion in '/home/jvican/.bloop/zsh/_bloop'
Installed Bash completion in '/home/jvican/.bloop/bash/bloop'
Installed systemd service in '/home/jvican/.bloop/systemd/bloop.service'
                                                                                                                                                                                                             
jvican in /data/rw/code/scala/loop                                                                                                                                                                [15:15:16] 
> $ python3 /data/rw/code/scala/loop/bin/install.py --dest ~/.bloop --nailgun 615737a9408654744bc066fa6c3cb938f3f139c6 --version 1.0.0-M9                                    [±tues-ticket/337]
Installed bloop server in '/home/jvican/.bloop/blp-server'
Installed bloop client in '/home/jvican/.bloop/bloop'
Installed zsh completion in '/home/jvican/.bloop/zsh/_bloop'
Installed Bash completion in '/home/jvican/.bloop/bash/bloop'
Couldn't download https://raw.githubusercontent.com/scalacenter/bloop/2e43fa5a/etc/systemd/bloop.service, please try again.
```
  
@tues Do you know what could be happening? I was surprised this was working at all in python2 because the bloop.service is not yet available in the GitHub repo URLs.

tues commented:
Sorry for the delay, new version:

* Changed `download_and_install_template` so it uses the same HTTP code as `download_and_install`
* Fixed file permissions
* Replaced `urlretrieve` with `urlopen` to make sure we don't ignore HTTP errors
* Rebased on top of `master`

Unfortunately it seems it's still failing on Windows. I'm not even sure it makes any sense to run the tests, I don't think they do anything with `install.py`? Maybe I should add `[DOC]`?
